cucumber-spring/src/test/java/io/cucumber/spring/TestTestContextAdaptorTest.java

```diff
@@ -43,12 +43,13 @@ void verifyNoMoroInteractions() {
 
     @Test
     void invokesAllLiveCycleHooks() throws Exception {
-        TestContextManager manager = new TestContextManager(SomeContextConfiguration.class);
-        TestContextAdaptor adaptor = new TestContextAdaptor(manager, singletonList(SomeContextConfiguration.class));
-        manager.registerTestExecutionListeners(listener);
-        InOrder inOrder = inOrder(listener);
+        TestContextAdaptor adaptor = new TestContextAdaptor(SomeContextConfiguration.class,
+            singletonList(SomeContextConfiguration.class));
 
         adaptor.start();
+
+        adaptor.getDelegate().registerTestExecutionListeners(listener);
+        InOrder inOrder = inOrder(listener);
```

Comment:
Assuming the race condition is caused by the concurrent creation of `TestContextManager`s some of these changes could be avoided by using a supplier. I.e:

```
new TestContextAdaptor(() -> manager, singletonList(SomeContextConfiguration.class))
```  

And then in the `TestContextManager` invoking this supplier inside a `synchronized (monitor)` block in the constructor.

Reply:
> And then in the TestContextManager invoking this supplier inside a synchronized (monitor) block in the constructor.

I think in this you mean in the `TestContextAdaptor` right? So in order to not introduce any changes in the unit test, I should just provide the `TestContextManager` instance inside the synchronized block of `TestContextAdaptor` block right. If that's correct, I will try that tomorrow to see how :D

Reply:
Yes! Ultimately I'd like to preserve the adaptor pattern. Having non-null final fields all the time is kinda nice.

Reply:
I've just updated the constructor of `TestContextAdaptor` to receive the Supplier of `TestContextManager` instead, so that the unit test for cucumber-spring has passed